src/UartSerial.cpp

```diff
@@ -92,48 +122,108 @@ void UartSerial::begin(uint32_t speed_, size_t config_)
   
   // Flush all current i/o data before enabling the new configuration
-  tcflush(_fd, TCIOFLUSH);
+  if (tcflush(_serial_file_descriptor, TCIOFLUSH) != 0)
+  {
+    perror("UartSerial::begin - unable to flush non-transmitted output data");
+    cleanup_serial_file_descriptor();
+    return;
+  }
 
   // Enable new term configuration
-  if(tcsetattr(_fd, TCSANOW, &_tio_config) < 0)
+  if(tcsetattr(_serial_file_descriptor, TCSANOW, &_tio_config) < 0)
   {
     perror("UartSerial::begin - Unable to set term attributes");
-    exit(-1);
+    cleanup_serial_file_descriptor();
+    return;
   }
 }
 
 void UartSerial::end()
 {
-  if(_fd != -1){
-    flush();
-    // Restore the original settings
-    tcsetattr(_fd, TCSANOW, &_tio_original_config);
-    close(_fd);
-    _fd = -1;
+  if(_serial_file_descriptor == -1) { return; }
+  
+  flush();
+
+  // Shut down the poll() thread
+  _polling = false;
+  _poll_thread.join();
+
+  // Restore the original settings
+  if (tcsetattr(_serial_file_descriptor, TCSANOW, &_tio_original_config))
+  {
+    perror("UartSerial::end - Unable to restore term attributes");
   }
+  cleanup_serial_file_descriptor();
 }
 
 void UartSerial::flush()
 {
-  tcdrain(_fd);
+  if(_serial_file_descriptor == -1) { return; }
+
+  if (tcdrain(_serial_file_descriptor) != 0)
+  {
+    perror("UartSerial::flush - tcdrain failed");
+  }
 }
 
 int UartSerial::read()
 {
-  int buffer[1];
-  int res = ::read(_fd, buffer, 1);
+  if(_serial_file_descriptor == -1) { return -1; }
+
+  uint8_t buffer;
+  int res = ::read(_serial_file_descriptor, &buffer, 1);
   if (res == 0) 
     return -1;
   else
-    return buffer[0];
+    return (int)buffer;
 }
 
 void UartSerial::write(uint8_t byte_)
 {
-  ::write(_fd, &byte_, 1);
+  if(_serial_file_descriptor == -1) { return; }
+
+  ::write(_serial_file_descriptor, &byte_, 1);
+}
+
+void UartSerial::registerSerialEventCallback (serialEvent bytes_available_, void *context_)
+{
+  if(_serial_file_descriptor == -1) { return; }
+
+  // For now only allow a single callback to be assigned
+  if(_bytes_available_callback == nullptr)
+  {
+    _bytes_available_callback = bytes_available_;
+    _callback_context = context_;
+
+    // Setup poll struct
+    _polling_file_descriptor.fd = _serial_file_descriptor;
+    _polling_file_descriptor.events = POLLIN;
+
+    _polling = true;
+    _poll_thread = std::thread(&UartSerial::poll_for_serial_data, this);
+  }
+}
+
+void UartSerial::poll_for_serial_data()
+{
+  int pollrc = 0;
+  while(_polling)
+  {
+    pollrc = poll(&_polling_file_descriptor, 1, 5); // timeout after 5ms
+    if (pollrc > 0 && (_polling_file_descriptor.revents & POLLIN))
+    {
+      _bytes_available_callback(_callback_context);
+    }
+    else
+    {
+      // Release control back to the CPU
+      this_thread::sleep_for(std::chrono::seconds(0));
+    }
+  }
 }
 
-void UartSerial::registerSerialEventCallback (serialEvent bytes_available_)
+void UartSerial::cleanup_serial_file_descriptor()
```

Comment:
int variable_names;
void functionNames (void);
void(*callbackVariableNames)(void);

But this is certainly not important and I feel douchy even mentioning it...